librz/analysis/var_global.c

```diff
@@ -0,0 +1,310 @@
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byaddr_at(analysis, addr);
+	if (!glob) {
+		return false;
+	}
+	// We need to delete RBTree first because ht_pp_delete will free its member
+	bool deleted = rz_rbtree_delete(&analysis->global_var_tree, &glob->addr, global_var_node_cmp, NULL, NULL, NULL);
+	return deleted ? ht_pp_delete(analysis->ht_global_var, glob->name) : deleted;
+}
+
+/**
+ * \brief Same as rz_analysis_var_global_delete_byname in the address
+ * 
+ * \param analysis RzAnalysis
+ * \param addr Global Variable address
+ * \return true if succeed
+ */
+RZ_API bool rz_analysis_var_global_delete_byaddr_in(RzAnalysis *analysis, ut64 addr) {
+	rz_return_val_if_fail(analysis, false);
+	RzAnalysisVarGlobal *glob = rz_analysis_var_global_get_byaddr_in(analysis, addr);
+	if (!glob) {
+		return false;
+	}
+	// We need to delete RBTree first because ht_pp_delete will free its member
+	bool deleted = rz_rbtree_delete(&analysis->global_var_tree, &glob->addr, global_var_node_cmp, NULL, NULL, NULL);
+	return deleted ? ht_pp_delete(analysis->ht_global_var, glob->name) : deleted;
+}
+
+/**
+ * \brief Get the instance of global variable by its name
+ * 
+ * \param analysis RzAnalysis
+ * \param name Global variable name
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_BORROW RzAnalysisVarGlobal *rz_analysis_var_global_get_byname(RzAnalysis *analysis, RZ_NONNULL const char *name) {
+	rz_return_val_if_fail(analysis && name, NULL);
+	return (RzAnalysisVarGlobal *)ht_pp_find(analysis->ht_global_var, name, NULL);
+}
+
+struct list_addr {
+	RzList /* <RzAnalysisVarGlobal> */ *list;
+	ut64 addr;
+};
+
+/**
+ * \brief Get the instance of global variable at the address
+ * 
+ * \param analysis RzAnalysis
+ * \param addr Global variable address
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_BORROW RzAnalysisVarGlobal *rz_analysis_var_global_get_byaddr_at(RzAnalysis *analysis, ut64 addr) {
+	rz_return_val_if_fail(analysis, NULL);
+
+	RBNode *node = rz_rbtree_find(analysis->global_var_tree, &addr, global_var_node_cmp, NULL);
+	if (!node) {
+		return NULL;
+	}
+	RzAnalysisVarGlobal *var = (RzAnalysisVarGlobal *)container_of(node, RzAnalysisVarGlobal, rb);
+	if (!var) {
+		return NULL;
+	}
+	return var;
+}
+
+/**
+ * \brief Get the instance of global variable contains the address
+ * 
+ * \param analysis RzAnalysis
+ * \param addr Global variable address
+ * \return RzAnalysisVarGlobal *
+ */
+RZ_API RZ_BORROW RzAnalysisVarGlobal *rz_analysis_var_global_get_byaddr_in(RzAnalysis *analysis, ut64 addr) {
+	rz_return_val_if_fail(analysis, NULL);
+
+	RBNode *node = rz_rbtree_upper_bound(analysis->global_var_tree, &addr, global_var_node_cmp, NULL);
+	if (!node) {
+		return NULL;
+	}
+	RzAnalysisVarGlobal *var = (RzAnalysisVarGlobal *)container_of(node, RzAnalysisVarGlobal, rb);
+	if (!var) {
+		return NULL;
+	}
+	ut64 size = rz_type_db_get_bitsize(analysis->typedb, var->type) / 8;
+	if (addr > (var->addr + size - 1) || addr < var->addr) {
```

Comment:
This check is correct but it can be simplified a bit:
```suggestion
	if (addr >= var->addr + size) {
```
`addr < var->addr` is already ensured by `rz_rbtree_upper_bound()`.